docs/contributing/git-style.rst

```diff
@@ -46,3 +46,98 @@ features or plugins to streamline this process (e.g.,
 `magit <https://magit.vc/>`_). Git also provides ``git commit --fixup`` and
 ``git rebase --autosquash`` which can streamline this process.
 
+It is also possible to ask GitHub to squash your commits for you. This is the
+"squash and merge" option when merging your PR.
+
+Commit Message Style
+====================
+
+Commit messages are part of the repository in the same way as the files in a
+repository are. And as developers may use ``grep`` to orient themselves around
+the repository, developers may also use ``git log`` and ``git blame`` to
+understand more of the history of a file and its context. So, it pays to be
+mindful of the content and structure of commits and their messages.
+
+Commit messages should be written with your reviewers in mind as well as future
+developers (or generally the codebase itself). Changes to the repository are
+already included in a commit so use your commit message to focus on things that
+would not readily be apparent from just looking at the textual changes.
+
+Here are some examples of useful things to note for your reviewers and future
+developers:
+
+- Why was this change made?
+
+- What plausible alternative implementations did you consider but rule out?
+
+- What general thing or abstraction is this change in furtherance of?
+
+- If this change fixes a bug, what was the original bug or its symptoms?
+
+- If you did some additional changes along the way to facilitate the main
+  thrust of your work, explain the rational behind any non-obvious but required
+  changes.
+
```

Comment:
I think this list should also include notable API changes. That makes it much easier for someone to understand when a commit may affect external code or other code they are working on.